boxtree/distributed/calculation.py

```diff
@@ -0,0 +1,417 @@
+    """
+    def __init__(self, context, comm, global_traversal,
+                 communicate_mpoles_via_allreduce=False):
+        self.context = context
+        self.comm = comm
+        self.global_traversal = global_traversal
+        self.communicate_mpoles_via_allreduce = communicate_mpoles_via_allreduce
+
+    def distribute_source_weights(self, src_weight_vecs, src_idx_all_ranks):
+        mpi_rank = self.comm.Get_rank()
+        mpi_size = self.comm.Get_size()
+
+        if mpi_rank == 0:
+            distribute_weight_req = []
+            local_src_weight_vecs = np.empty((mpi_size,), dtype=object)
+
+            for irank in range(mpi_size):
+                local_src_weight_vecs[irank] = [
+                    source_weights[src_idx_all_ranks[irank]]
+                    for source_weights in src_weight_vecs]
+
+                if irank != 0:
+                    distribute_weight_req.append(self.comm.isend(
+                        local_src_weight_vecs[irank], dest=irank,
+                        tag=MPITags["DIST_WEIGHT"]
+                    ))
+
+            MPI.Request.Waitall(distribute_weight_req)
+            local_src_weight_vecs = local_src_weight_vecs[0]
+        else:
+            local_src_weight_vecs = self.comm.recv(
+                source=0, tag=MPITags["DIST_WEIGHT"])
+
+        return local_src_weight_vecs
+
+    def gather_potential_results(self, potentials, tgt_idx_all_ranks):
+        mpi_rank = self.comm.Get_rank()
+        mpi_size = self.comm.Get_size()
+
+        from boxtree.distributed import dtype_to_mpi
+        potentials_mpi_type = dtype_to_mpi(potentials.dtype)
+
+        if mpi_rank == 0:
+            # The root rank received calculated potentials from all worker ranks
+            potentials_all_ranks = np.empty((mpi_size,), dtype=object)
+            potentials_all_ranks[0] = potentials
+
+            recv_reqs = []
+
+            for irank in range(1, mpi_size):
+                potentials_all_ranks[irank] = np.empty(
+                    tgt_idx_all_ranks[irank].shape, dtype=potentials.dtype)
+
+                recv_reqs.append(
+                    self.comm.Irecv(
+                        [potentials_all_ranks[irank], potentials_mpi_type],
+                        source=irank, tag=MPITags["GATHER_POTENTIALS"]))
+
+            MPI.Request.Waitall(recv_reqs)
+
+            # Assemble potentials from worker ranks together on the root rank
+            potentials = np.empty(
+                self.global_traversal.tree.ntargets, dtype=potentials.dtype)
+
+            for irank in range(mpi_size):
+                potentials[tgt_idx_all_ranks[irank]] = potentials_all_ranks[irank]
+        else:
+            # Worker ranks send calculated potentials to the root rank
+            self.comm.Send([potentials, potentials_mpi_type],
+                           dest=0, tag=MPITags["GATHER_POTENTIALS"])
+
+        return potentials
+
+    def slice_mpoles(self, mpoles, slice_indices):
```

Comment:
Maybe add underscores to the front of the name of internal methods? I know that other places in boxtree also don't do this, but I've started viewing it as good practice.